### PR TITLE
Improve the editor theme

### PR DESCRIPTION
**Pull request by Calinou**
The editor theme now makes use of rounded corners and less borders to follow modern visual trends. **The corner radius and border width are configurable in the Editor Settings**, so you can revert to the old appearance by setting corner radius to 0 and border width to 1.

The default theme's colors were also tweaked to make the blue hue more subtle (similar to the Arc theme, which was removed as a consequence). The Alien theme was replaced by a Breeze Dark theme, which should blend in well with the KDE theme.

The new editor theme will allow distinguishing Godot 4.0 easily from Godot 3.2, even if the version number is not visible on screenshots/videos. This is important for tutorials and the like :slightly_smiling_face: 

## TODO

- [ ] Increase contrast (make the main background darker) by reduz's request.
  - In a future PR, increase saturation for icon colors. This will require modifying all the SVGs and tweaking the light theme icon conversion code accordingly.
    - **Edit:** Done in https://github.com/godotengine/godot/pull/48644.
- [ ] Make corner radii more consistent across nodes, such as the top of PopupMenus.
  - [ ] Consider making LineEdits/TextEdits not use rounded corners to distinguish them more easily from buttons.
- [x] Fix some 2D/3D editor buttons not being marked as flat when they should be (following the ToolButton removal).
- [x] Fix PopupMenu hover styling so the highlight line spans the whole width of the popup.
  - Done by https://github.com/godotengine/godot/pull/45704.
- [ ] Improve tab/subtab styling in windows such as the project/editor settings. I'm not sure how to improve it, so ideas are welcome.
  - [ ] This can be done by changing the background color of subwindows to be darker. It might not always be a good idea, so we can try that in a future PR.

## Preview

*Click to view at full resolution.*

***Note:** The screenshots below are outdated and don't reflect the version of the pull request that was merged.*

### Dark theme

Editor

Editor Settings

PopupMenu

### Light theme

Editor

### CODE REVIEW

**Comment by pycbouh** [comment]
There are some weird artifacts due to the new border styling, namely in tabs and window headers:

image
image

**Comment by Calinou** [comment]
> There are some weird artifacts due to the new border styling, namely in tabs and window headers:

For the first case, this is due to antialiasing blurring out borders as long as one of the corners has a radius greater than zero. It regressed sometime around 3.2.2 or 3.2.3.

For the second case, this is due to me not setting the bottom corner radii of the window title to 0. I'll look into fixing that.

**Comment by HungryProton** [comment]
I'm all in for an editor theme overhaul, but I think having consistent margins is more important than putting rounded corners everywhere. (Don't get me wrong, I like rounded corners).

One case in particular is in this screenshot, having the highlighted part stuck to the text is quite jarring in my opinion.
image

It would look better if the hover style took the entire width, without the rounded corners, kind of like this: (ignore the colors)
image

**Comment by Calinou** [comment]
> One case in particular is in this screenshot, having the highlighted part stuck to the text is quite jarring in my opinion.

It used to look like the second screenshot you posted, but it regressed due to the PopupMenu refactor.

**Comment by barbaros83** [comment]
i think the inspector could also use an update, it sometimes hard to distinguish where an expanded tab starts and ends

**Comment by Calinou** [comment]
> i think the inspector could also use an update, it sometimes hard to distinguish where an expanded tab starts and ends

This is something reduz also mentioned, but I'd prefer to tackle it in a future PR.

**Comment by EricEzaM** [comment]
@HungryProton

It is I, PopupMenu person. How's this?

* removed horizontal margin
* kept vertical margin
* added horizontal padding for items.

image

**Comment by HungryProton** [comment]
@EricEzaM 
Looks a lot better like this in my opinion! (I'm no PopupMenu authority though)

**Comment by EricEzaM** [comment]
I agree. Much more modern. I will submit a PR with these changes.

**Comment by EricEzaM** [comment]
Overall on the theme though, I must say I quite dislike the mix of sharp and rounded corners. I think one should be chosen. Sharp edges is by far more common in the editor currently and it looks pretty clean. 

The rounded edges are just sort of there randomly... on PopupMenu selection (why?), on SceneTree selection, on the search bar, on the _unselected_ tabs but not on the selected tabs (???). Just clashes imo.

**Comment by Calinou** [comment]
I made the selected tabs have rounded corners (for consistency with unselected tabs) and fixed the corner bug between single-window mode window titles and bodies:

image

**Comment by fabriceci** [comment]
I found the rounded selected tab looks like there is a problem, plus the left border stop abruptly. In Blender you have some elements rounded others not, it's hard the keep coherence.

Capture d’écran 2021-02-04 à 20 20 48
 
Maybe this PR should be split in multiple little PR step by step (like #45704) , start by steps where there is a consensus (maybe color? decrease the padding? etc.).

**Comment by Calinou** [comment]
> Maybe this PR should be split in multiple little PR step by step (like #45704) , start by steps where there is a consensus (maybe color? decrease the padding? etc.).

Unfortunately, things tend to depend on each other when you do UI design, so this is not really a possibility.

**Comment by fabriceci** [comment]
> I can fix this by making the tabs' top corners non-rounded, let's give it a try.

@Calinou this is way way much better!

- Still 2 minor remarks

1. The arrows should be aligned (same padding right and same arrow (size?)):

Screenshot 2021-02-05 at 10 39 25

2. The icon of the button should be centered or the button's size reduced.
Screenshot 2021-02-05 at 10 39 36

In the subjective area, I had suggestions:

1. The toolbar seems to me more airy/cleaner in the current version than the new one, mainly due to the lack of separators.

Screenshot 2021-02-05 at 10 39 49
Screenshot 2021-02-05 at 17 16 02

2.  In the settings panel there is not left padding before the text, I don't know if it's customizable:
Screenshot 2021-02-05 at 17 20 13
Screenshot 2021-02-05 at 17 20 16

3. In the settings panel, in the treeview, the items are higher than the headers. If again it's customizable, it will be great to decrease de padding up/down, to be able to see more element.

**Comment by Calinou** [comment]
> The arrows should be aligned (same padding right and same arrow (size?)):

That issue was already present before the editor theme redesign. I'd prefer tackling stuff like this in a future PR.


> 2. The icon of the button should be centered or the button's size reduced.

This is a "ghost" dropdown which should be removed entirely until an OpenGL renderer is implemented in the `master` branch.

**Comment by Calinou** [comment]
I've added outlines below LineEdits and TextEdits to make them easier to notice and distinguish from OptionButtons:

image

**Comment by Riteo** [comment]
I don't know, the fact that the line is under the rounded rectangle makes it look kind of weird IMO, why not putting it on top of the field?

**Comment by Riteo** [comment]
Here's my opinion:
Usually, one selected tab has its small little space for its content.
tabs
This isn't the case for sub-windows though. There, the active background color is, for some reason, the main color, except around the tabs, which creates this kind of ugly square outline:
weird outline
This could be resolved by making the background color consistent with the main window.
Also, I think that it would look better if the tabs were rounded. I made a little mock-up explaining what I mean:
mock up

Also, as I noted earlier, there are some artifacts with text fields inside sub-windows, which don't appen at all in the main window.
Main window:
main window TextEdit

Sub-window:
artifacts

I have no idea if these issues are actually only present in sub-windows, but AFAICT it looks like it.

**Comment by Calinou** [comment]
> Edit: I just now noticed how actually they're more like an outline, since they follow the curves of the fields, but for some reason in the editor preferences this doesn't apply, making it look like there's a normal line under them, and that's what I was talking about.

This is because I made the bottom corners non-rounded. Otherwise, this kind of line always ends up looking weird.

**Comment by Riteo** [comment]
> See above. There are just as many people disliking rounded tab highlights as people liking them. I can't please everyone slightly_smiling_face
(And no, I won't add an option for such a small difference.)

Oh, I forgot about that, no worries then. I prefer it this way since it's more consistent with the rounded style IMO, but it's okay too if there are square tabs.

>This is because I made the bottom corners non-rounded. Otherwise, this kind of line always ends up looking weird.

But in the main window they are still rounded, is this intended behaviour?

Any idea about the weird artifacts in the text fields in the sub window? I guess that's because the outline isn't rounded anymore. Ignore the circled(or squared? lol) one, look at the corners.
artifacts

**Comment by Calinou** [comment]
I think this is because those LineEdits have a dark background coming from the sectioned inspector itself. The LineEdit's top corners are rounded, but the sectioned inspector's corners aren't, which makes them visible.

**Comment by Riteo** [comment]
> sectioned inspector
What do you mean for sectioned inspector?

**Comment by Calinou** [comment]
> What do you mean for sectioned inspector?

The Project Settings and Editor Settings use the sectioned inspector.

**Comment by Riteo** [comment]
@Calinou oh, didn't know about that class, thanks for letting me know. I haven't touched a lot of GUI code if not maybe `Window`, since I worked a bit on the X11 `DisplayServer` before nvidia broke it badly.

**Comment by gongpha** [comment]
These ColorPicker buttons make these rows thinner than others.
107067494-a5164480-67df-11eb-89c0-118d3ada77f7

**Comment by Calinou** [comment]
@gongpha This is a regression from the TextServer changes; it's unrelated to this pull request.

**Comment by gongpha** [comment]
> > These ColorPicker buttons make these rows thinner than others.
> 
> This is a regression from the [TextServer](https://godotengine.org/article/complex-text-layouts-progress-report-3) changes. It's unrelated to this pull request.

Oh, I thought it happened because of the flat setting.

And I think it would be well. If the controls' corners had more rounder, it would look like Blender UI and modest webs (like Github, now the web look roundly).

**Comment by Calinou** [comment]
> And I think it would be well. If the controls' corners had more rounder, it would look like Blender UI and modest webs (like Github, now the web look roundly).

I don't want to make corners look too rounded either.

**Comment by alexfreyre** [comment]
I like a lot this PR and I want to share some ideas.

I propose this little change, just let's make the tabs take up all the space around. Why do we leave that space?

tabs_1
tabs_2

**Comment by alexfreyre** [comment]
I'm a bit worried about one thing, although this is an update of the current theme, I think it is better to dim in the new theme the areas that are not being used at the moment or not to point them.

While the user tries to concentrate on his work in the viewport, the editor should not attract more attention than the work itself.

I think the interface should be a little more neutral and simple.

For example, while I want to work in the viewport, all these elements catch my attention 
new_design

**Comment by Calinou** [comment]
> I propose this little change, just let's make the tabs take up all the space around. Why do we leave that space?

This should be handled in a separate PR.

**Comment by alexfreyre** [comment]
I want to share a few design insights that I think are helpful for the problems faced by this PR.

Note that saturated colors and light values ​​attract attention, putting a lot of elements complicates things.

less is more, there are subtle ways to create groups and differentiate things, to separate elements it is not always necessary to use borders, we can use:

- zones of different values
the dark tones recede, the light ones advance
therefore a scale of values ​​indicates an overlap of elements

- certain colors as an emphasis
. colors attract more attention than values
. each color affects the perception of the other
(for these reasons, many software adopt a medium gray interface with some accent colors)
. each color can designate a group
. each color and its values ​​(from white to black) can designate a group and subgroups (harmonically)
. each color and its mixture with another color (tones) create a color family
. each color and its complementary is a very dangerous mixture, it claims a lot of attention (red-green, yellow-purple, blue-orange)
. non-complementary colors are analogous and their mixture is harmonious (red-orange, orange-yellow, blue-green, green-yellow, blue-purple, purple-red)

I hope this small contribution is useful. :)

**Comment by interpol-kun** [comment]
What about margins? I think that current look looks like bug. Maybe a little bit of space before the text would be great. 
image
image

**Comment by Calinou** [comment]
> What about margins? I think that current highlighted panel text looks like a bug. Maybe a little bit of space before the text would be great.

I'd prefer fixing this in a separate pull request since this needs to be fixed/improved in the individual control nodes.

**Comment by alexfreyre** [comment]
Look @Calinou I have found this example just now, from blender 2.92 new features page. This image illustrates what I was talking about before, all zones are clearly separated by different values, looks elegant, simple, it works and solves the problem.

While your solutions is good! I am bringing in this case a reference to take into account. As you can see dark values goes back and are in the example the big zones, higher values goes up and for little zones. Personally I see this solution simple and effective.   

Probably we can use borders too and/or highlight the area in which the user has the mouse over and everything else can be a little neutral.

I like a lot the UI of Godot, more than unreal, unity and a lot more of software. It can be better and it needs work, lots of things are being added for 4.0 and we need to give some love to the UI. I agree with a design that could be like the one below (or not), but I see on this for example rounded corners and a very clean as a good solution. 

image

**Comment by jasperbrooks79** [comment]
I'd consider adding, so all icons, are 1 color, to give it more consistent look . . at least, as an option . . . maybe, as toggle, ' monochrome icons ' . . I mean, in node inspector . . :O would give it that ' classic ' application look . .

I love everything I see here, can't wait, Thanks . . 10 / 10 <3

**Comment by pycbouh** [comment]
@jasperbrooks79 You will be able to make icons look monochrome in 4.0. See https://github.com/godotengine/godot/pull/45924

**Comment by jasperbrooks79** [comment]
Ok, a few thoughts . .
Skærmbillede (1466)

Basically, all these things can be ' fixed ', if we could ' set ' the rounded value, for the different parts, so users could ' tweak ', that would give best of both worlds, a new look, and an easy way, to make it ' custom ', for almost all users . . :OO I really like this entire proposal, I hope all these things will be added, as customization options, and when that has been done, the core - team can make some themes, together, after this becomes the ' main theme ' . . With a theme editor, and all these things as options, making it ' just right ', be a breeze <3 <3 . .

Skærmbillede (1467)

Skærmbillede (1468)

ps. Really like the slightly rounded look, around icons :OO again, since some users will want the ' old ' look, some, are different . . it'd be a good idea, if this could be a customization toggle, ie. 1 pixel roundedness, 2 pixel roundedness, or I think, 0 for the old look :OO 

All these features, they all feel like an improvement to me, IF they were customization options, 1. it'd be nice to have a new look, for Godot 4, a bit different, polished . . 2. users could set them, to their favorites, assuming any-one, would want to change this . .

To keep it short, Thank you . . . <3 . .

**Comment by Xrayez** [comment]
> image

I also think that the accent is too bright, that's the first thing which caught my eye with the new theme, especially with panel outlines (unless this was already fixed).

**Comment by Riteo** [comment]
> it'd be a good idea, if this could be a customization toggle, ie. 1 pixel roundedness, 2 pixel roundedness, or I think, 0 for the old look :OO

This PR already adds an editor setting for that :)

**Comment by Calinou** [comment]
I made some changes to prepare for merging:

- Make the active tab highlights less prominent.
- Make Trees have a slightly darker background to make them easier to distinguish from LineEdits/TextEdits.
- Don't use rounded corners for PopupMenu hover highlights since they now fully extend to the PopupMenu's corners.

image

**Comment by Ansraer** [comment]
Looking at all the screenshots that have been posted so far I have to say that I definitely prefer the tabs with rounded top corners. imo the rounded corners make it easier to see where one tab ends and the next begins.

Have you ever considered removing the padding around the entire editor?

**Comment by aaronfranke** [comment]
@Ansraer He did consider it #40287

File: editor/plugins/script_editor_plugin.cpp [hunk 1]
```diff
@@ -1842,11 +1841,7 @@ void ScriptEditor::_update_script_colors() {
 
 		script_list->set_item_custom_bg_color(i, Color(0, 0, 0, 0));
 
-		bool current = tab_container->get_current_tab() == c;
-		if (current && highlight_current) {
-			script_list->set_item_custom_bg_color(i, EditorSettings::get_singleton()->get("text_editor/script_list/current_script_background_color"));
```
Comment (akien-mga) [inline editor/plugins/script_editor_plugin.cpp:1847]: How does the script list look like with this removed? Highlighting the current script seems pretty important to me to know what you're editing.

**Comment by junkajii** [comment]
About the Toggle Split Mode button I think better without background for consistency

toggle-split-mode-gd4

File: editor/plugins/script_editor_plugin.cpp (hunk 1 above)
Comment (Calinou) [reply editor/plugins/script_editor_plugin.cpp:1847]: The script highlight was actually being done twice – the code being present there was a bug all along :slightly_smiling_face: 

Removing this code makes the highlight only appear once, as the script editor plugin already applies some highlight on its own.

**Review by akien-mga — APPROVED** [review]
Let's give this a spin and address additional feedback in follow-up issues/PRs once we've gotten used to the changes.

**Comment by akien-mga** [comment]
Thanks!

**Comment by SilencedPerson** [comment]
On twitter i saw mockup that removed the outer borders, could we instead get rid of the inner borders to remove bit of visual clutter but keep the visible separation of the information? Here is quick and dirty mockup:
image

**Comment by pycbouh** [comment]
@SilencedPerson If you are talking about the removal of paddings inside of the docks, it was the original idea (by jfons, I think?) which groud take one step further to remove gaps as well. It's a good idea, but it would require to change the actual layout of every dock (so it's not just a theme change), which makes it harder to try as an experiment.

Feel free to open a proposal, I guess. It would be preferable to discuss it there rather than in a closed PR.

**Pull request merged**